### ISSUE TITLE
Unbound variable in try/except..as/finally block

### ISSUE THREAD

**Issue by ozars** [microsoft/pyright#2402]
**Describe the bug**

An initialized variable is inferred as unbound when ```except...as``` with the same variable name is involved.

**To Reproduce**
```python
from __future__ import annotations

exc = None
# exc: RuntimeError | None = None

try:
    f = open('somefile')
except RuntimeError as exc:
    reveal_type(exc)
finally:
    reveal_type(exc)
```

```console
  /home/omer/tmp/pyright/deneme.py:9:17 - info: Type of "exc" is "RuntimeError"
  /home/omer/tmp/pyright/deneme.py:11:17 - info: Type of "exc" is "Unbound | None"
  /home/omer/tmp/pyright/deneme.py:11:17 - error: "exc" is possibly unbound (reportUnboundVariable)
```

(same issue happens with the commented alternative as well)

**Expected behavior**

Second reveal_type returns `RuntimeError | None`. This works:

```python
from __future__ import annotations

exc = None
#  exc: RuntimeError | None = None

try:
    f = open('somefile')
except RuntimeError as e:
    exc = e
finally:
    reveal_type(exc)
```

**VS Code extension or command-line**

pyright 1.1.177 cli

**Comment by erictraut** [comment]
I think the behavior is correct. It's due to a rather arcane scoping rule employed in the Python runtime. If an except clause uses a target variable to capture the expression, that variable is unbound when execution leaves the except scope.

For example, run the following code:
```python
exc = None
print(exc)

try:
    print(exc)
    raise ValueError()
except ValueError as exc:
    pass
finally:
    print(exc)
```

The output is:
```
None
None
NameError: name 'exc' is not defined
```
